ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockHelixCluster.java

```diff
@@ -150,29 +150,59 @@ void bringAllInstancesDown() {
   }
 
   /**
+   * Set the state of a partition
+   * @param partition partition for which state needs to be updated
+   * @param partitionState {@link PartitionState} that needs to be set
+   */
+  void setPartitionState(String partition, PartitionState partitionState) {
+    for (MockHelixAdmin helixAdmin : helixAdmins.values()) {
+      helixAdmin.setPartitionState(partition, partitionState);
+    }
+  }
+
+  /**
    * @return the set of all partitions in this cluster.
    */
-  Set<String> getPartitions() {
-    Set<String> partitions = new HashSet<>();
+  Set<String> getAllPartitions() {
+    Set<String> partitions = null;
     for (MockHelixAdmin helixAdmin : helixAdmins.values()) {
-      partitions.addAll(helixAdmin.getPartitions());
+      if (partitions == null) {
+        partitions = new HashSet<>(helixAdmin.getPartitions());
+      } else {
+        partitions.retainAll(helixAdmin.getPartitions());
+      }
     }
     return partitions;
   }
 
   /**
    * @return the set of all partitions in this cluster that are up.
    */
-  Set<String> getUpPartitions() {
-    Set<String> upPartitions = null;
+  Set<String> getWritablePartitions() {
+    Set<String> writablePartitions = null;
+    for (MockHelixAdmin helixAdmin : helixAdmins.values()) {
+      if (writablePartitions == null) {
+        writablePartitions = new HashSet<>(helixAdmin.getWritablePartitions());
+      } else {
+        writablePartitions.retainAll(helixAdmin.getWritablePartitions());
```

Comment:
same question here. why wouldn't this return a union in case the partitions are different in different datacenters?